redhat-appstudio/service-provider-integration-operator
# Ensure that secret type kubernetes.io/dockerconfigjson is properly generated for the Quay image provider and can be used as Deployment's imagePullSecret

Signed-off-by: Pavol Baran <pbaran@redhat.com>

### What does this PR do?
When a user specifies secret type as kubernetes.io/dockerconfigjson in binding, this PR ensures a proper secret is created that can be used as an imagepullsecret in podspec.

### Screenshot/screencast of this PR

binding:
![image](https://user-images.githubusercontent.com/73115616/170031703-5551b160-ed9e-4239-b66f-5c618d2b392a.png)
secret:
![image](https://user-images.githubusercontent.com/73115616/170031909-a98709ba-602f-4113-b111-dd0204f53125.png)
successfully pulled image:
![image](https://user-images.githubusercontent.com/73115616/170036183-9370e6bf-268a-4727-8b77-16d154900192.png)


### What issues does this PR fix or reference?

fixes: https://issues.redhat.com/browse/SVPI-115

### How to test this PR?

1. deploy operator with changes; wait for all components to be running
` make deploy_minikube SPIO_IMG=quay.io/pbaran/spio:SVPI-115-DOCKERCONFIG`

2. create a token binding that has a secret type: kubernetes.io/dockerconfigjson and repo URL that specifies a private quay image that can be able pulled to a pod, such as: 
`kubectl apply -f samples/binding-dockerconfigjson.yaml`

3. go through the OAuth flow

4. check that dockerconfigjson secret was created and create a Pod that uses it as imagepullsecret:
`kubectl apply -f test-pod`
where test-pod looks like: 
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: private-reg-pod
spec:
  containers:
  - name: private-reg-container
    image: <private_image>
  imagePullSecrets:
  - name: test-secret-dockerconfigjson
```
5. check that the pod has successfully pulled the image
